scripts/check_changeset_completeness.ts

```diff
@@ -2,6 +2,73 @@ import getReleasePlan from '@changesets/get-release-plan';
 import { GitClient } from './components/git_client.js';
 import { readPackageJson } from './components/package-json/package_json.js';
 import { EOL } from 'os';
+// eslint-disable-next-line import/no-extraneous-dependencies
+import { ReleasePlan } from '@changesets/types';
+
+const getModifiedPackages = (changedFiles: string[]): Set<string> => {
+  const modifiedPackageDirs = new Set<string>();
+
+  changedFiles
+    .filter(
+      (changedFile) =>
+        changedFile.startsWith('packages/') && !changedFile.endsWith('test.ts')
+    )
+    .forEach((changedPackageFile) => {
+      modifiedPackageDirs.add(
+        changedPackageFile.split('/').slice(0, 2).join('/')
+      );
+    });
+  return modifiedPackageDirs;
+};
+
+const checkBackendDependenciesVersion = (releasePlan: ReleasePlan) => {
+  const backendDependencies: string[] = [
+    '@aws-amplify/backend-auth',
+    '@aws-amplify/backend-data',
+    '@aws-amplify/backend-function',
+    '@aws-amplify/backend-storage',
+  ];
+  const backendName: string = '@aws-amplify/backend';
+  const versionBumpOfWrongKind: string[] = [];
+  let backendMaxVersionType: string = 'none';
+
+  for (const changeset of releasePlan.changesets) {
+    for (const release of changeset.releases) {
+      if (release.name === backendName) {
+        if (
+          release.type === 'major' ||
+          backendMaxVersionType === 'none' ||
+          (backendMaxVersionType === 'patch' && release.type === 'minor')
+        ) {
+          backendMaxVersionType = release.type;
+        }
+      }
+    }
+  }
+
+  for (const changeset of releasePlan.changesets) {
+    for (const release of changeset.releases) {
+      if (backendDependencies.includes(release.name)) {
+        if (
+          backendMaxVersionType !== release.type &&
+          (release.type === 'major' ||
+            backendMaxVersionType === 'none' ||
+            (backendMaxVersionType === 'patch' && release.type === 'minor'))
+        ) {
+          versionBumpOfWrongKind.push(release.name);
+        }
+      }
+    }
+  }
+
+  if (versionBumpOfWrongKind.length > 0) {
+    throw new Error(
+      `${backendName} has a version bump of a different kind of the following packages but is expected to have a version bump of the same kind:${EOL}${versionBumpOfWrongKind.join(
+        EOL
+      )}`
+    );
+  }
```

Comment:
This code is a bit hard to read.

Consider adding `const maxVersion = (...versionType: Array<string>): string => { // some logic to pick max }`


```suggestion
  const backendName: string = '@aws-amplify/backend';
  let effectiveBackendVersionType: string = 'none';
  let effectiveBackendDependenciesVersionType: string = 'none';

  for (const changeset of releasePlan.changesets) {
    for (const release of changeset.releases) {
      if (release.name === backendName) {
        effectiveBackendVersionType = maxVersion(effectiveBackendVersionType, release.type);
      }
    }
  }

  for (const changeset of releasePlan.changesets) {
    for (const release of changeset.releases) {
      if (backendDependencies.includes(release.name)) {
        effectiveBackendDependenciesVersionType = maxVersion(effectiveBackendDependenciesVersionType, release.type);
      }
    }
  }

  if (effectiveBackendVersionType !== effectiveBackendDependenciesVersionType) {
    throw new Error(
      `${backendName} has a version bump of a different kind of the following packages but is expected to have a version bump of the same kind:${EOL}${versionBumpOfWrongKind.join(
        EOL
      )}`
    );
  }
```

Reply:
You could also consider something like this.

Assuming `maxVersion(...versions)` and `findEffectiveVersion(releasePlan, packageName`

have something like that:
```
const backendVersion = findEffectiveVersion(releasePlan, '@aws-amplify/backend`);
const backendAuthVersion = findEffectiveVersion(releasePlan, '@aws-amplify/backend-auth`);
const backendDataVersion = findEffectiveVersion(releasePlan, '@aws-amplify/backend-data`);
...

if (backendVersion !== maxVersion(backendAuthVersion, backendDataVersion, ...)) { throw}
```
```

Reply:
And. don't worry about optimizing loops. These arrays have ~10-20 items. Ability to understand what code is doing is more important here.

Reply:
Would you prefer I remove the `versionBumpOfWrongKind` array as well?

Reply:
yes.

Reply:
And.. there's one interesting edge case here.

It might happen that `@aws-amplify/backend` will get version bump but all of `backend-X` will not. That should be a legit case.

I.e. the condition should be  `if(effectiveBackendVersionType < effectiveBackendDependenciesVersionType) { throw }` (this is pesudocode).

For that you might either consider either creating some kind of comparator function for `minor|major|patch|none`

or create enum with numerical values like `enum { MAJOR: 3, MINOR: 2... }`..

Reply:
enum sounds like the cleaner option. My original code was able to handle cases where `backend` got changes but `backend-X` didn't through this: `backendMaxVersionType !== release.type && (release.type === 'major' || backendMaxVersionType === 'none' || (backendMaxVersionType === 'patch' && release.type === 'minor'))`
but it is a fairly ugly looking set of code that is not intuitive to understand.